### PR TITLE
fix: wrap seed reference into react hooks

### PR DESCRIPTION
**Pull request by hanwencheng**
By using hooks: UI will notify and check if `valid` and `dataRef` existed.

In React component just get seedRef by 

```typescript
const [valid, seedRef, create, destory] = useSeedRef();
if(!valid){
  create(seedPhrase, password)
}
```

### CODE REVIEW

File: src/utils/seedRefHooks.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { useState } from 'react';
+
+import { decryptDataRef, destroyDataRef } from 'utils/native';
+
+type TryCreateFunc = (encryptedSeed: string, password: string) => Promise<void>;
+type TryDestroyFunc = () => Promise<void>;
+
+type SeedRefHooks = [number, boolean, TryCreateFunc, TryDestroyFunc];
```
Comment (sjeohp) [inline src/utils/seedRefHooks.ts:8]:
```suggestion
type SeedRefHooks = [number, boolean, TryCreateFunc, TryDestroyFunc, TryBrainWalletSign, TrySubstrateSign];
```

File: src/utils/seedRefHooks.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { useState } from 'react';
+
+import { decryptDataRef, destroyDataRef } from 'utils/native';
+
+type TryCreateFunc = (encryptedSeed: string, password: string) => Promise<void>;
+type TryDestroyFunc = () => Promise<void>;
+
```
Comment (sjeohp) [inline src/utils/seedRefHooks.ts:7]: type TrySign = (message: string) => Promise<string>;

Comment (sjeohp) [inline src/utils/seedRefHooks.ts:7]:
```suggestion
type TrySign = (message: string) => Promise<string>;

```

File: src/utils/seedRefHooks.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	// Destroy the decrypted seed. Must be called before this leaves scope or
+	// memory will leak.
+	const destroy: TryDestroyFunc = function () {
+		return destroyDataRef(dataRef).then(() => {
+			setValid(false);
+		});
+	};
+
+	return [dataRef, valid, create, destroy];
```
Comment (sjeohp) [inline src/utils/seedRefHooks.ts:30]:
```suggestion
    // Use the seed reference to sign a message. Will throw an error if
    // `destroy` has already been called or if `create` failed.
    const brainWalletSign: TrySign = function (message) {
        if (!valid) {
            // Seed reference was never created or was already destroyed.
            throw new Error('cannot sign with an invalid seed reference');
        }
        return EthkeyBridge.brainWalletSignWithRef(dataRef, message);
    }

    // Use a reference returned by decryptDataRef to sign a message
    const substrateSign: TrySign = (message) {
        if (!valid) {
            // Seed reference was never created or was already destroyed.
            throw new Error('cannot sign with an invalid seed reference');
        }
        return EthkeyBridge.substrateSignWithRef(dataRef, message);
    }
    
	return [dataRef, valid, create, destroy, brainWalletSign, substrateSign];
```

Comment (sjeohp) [reply src/utils/seedRefHooks.ts:30]: Signing functions should be included to make sure `valid` is always checked before calling.

**Comment by sjeohp** [comment]
Looks good.  I've added signing functions since we want them to be as tightly coupled as possible to `create` and `destroy`.

**Comment by sjeohp** [comment]
I have included this in #583 with some additional changes.

**Comment by hanwencheng** [comment]
Closed as included in #583

**Pull request closed without merging**